bunnymatic/bryantstreetstudios
# Add CSRF protection

problem
-----

As we move to using more Ajax and stuff, we'd like to have some CSRF protection.

Basically this is implemented by putting a token in the cookie and let any post request
from the frontend come back and send that information back to the server.  This allows
the server to know that the request came from the same origin and with the same request
channel.  this will prevent cross site hits to our post endpoints.

Duplicate and revamp of #380

Solution
------

* Hand roll CSRF protection.
* Add the csrf token to the meta tags
* add the csrf input to forms
* add the csrf input to the api when we call for instagram feed
* move instagram call to a post so we can protect it
* Move serveral specs to e2e because we need javascript working
* update specs
* Add capybara helper to fill in our markdown editor on the admin side
* limit Rack::Protection a bit - don't include RemoteToken which blows any links from external sites (like search engine results)

Tests
-----

in addition to rspec, I tested IRL using the following:

* build local assets `bin/vite build`
* run the server `bin/rackup`
* run ngrok `ngrok http http://localhost:9292`
* make a codepen that had a link to the ngrok site `<a target="_blank" href="https://whatever-my-ngrok-domain-was">here</a>`
* click the link to make sure it worked.
* try again after re-instating `remote-token` for rack protection in config.ru to verify that was the problem child

Also tested IRL updating form data via admin side which is where the CSRF stuff was supposed to be protecting us.
